chainimport/headers_import.go

```diff
@@ -0,0 +1,1816 @@
+	filterHeaderStore := s.options.TargetFilterHeaderStore
+	targetFilterHeader, err := filterHeaderStore.FetchHeaderByHeight(
+		height,
+	)
+	if err != nil {
+		return fmt.Errorf("failed to get filter header from target at "+
+			"height %d: %w", height, err)
+	}
+
+	// Compare filter headers from import and target sources.
+	sourceFilterHeaderHash := importSourceFilterHeader.FilterHash
+	targetFilterHeaderHash := *targetFilterHeader
+	if !sourceFilterHeaderHash.IsEqual(&targetFilterHeaderHash) {
+		return fmt.Errorf("filter header mismatch at height %d: "+
+			"source has %v but target has %v", height,
+			sourceFilterHeaderHash, targetFilterHeaderHash)
+	}
+
+	log.Debugf("Headers from %s (block) and %s (filter) verified at "+
+		"height %d", s.BlockHeadersImportSource.GetURI(),
+		s.FilterHeadersImportSource.GetURI(), height)
+
+	return nil
+}
+
+// processNewHeadersRegion imports headers from the specified region into the
+// target stores. This method handles the case where headers exist in the import
+// source but not in the target stores.
+func (s *HeadersImport) processNewHeadersRegion(region HeaderRegion,
+	result *ImportResult) error {
+
+	if !region.Exists {
+		return nil
+	}
+
+	log.Infof("Adding %d new headers (block and filter) from heights "+
+		"%d to %d", region.End-region.Start+1, region.Start, region.End)
+
+	err := s.appendNewHeaders(region.Start, region.End)
+	if err != nil {
+		return fmt.Errorf("failed to append new headers: %w", err)
+	}
+
+	result.ProcessedCount += int(region.End - region.Start + 1)
+	result.AddedCount += int(region.End - region.Start + 1)
+
+	return nil
+}
+
+// appendNewHeaders adds new headers from import source.
+func (s *HeadersImport) appendNewHeaders(startHeight, endHeight uint32) error {
+	// Get metadata from sources.
+	metadata, err := s.BlockHeadersImportSource.GetHeaderMetadata()
+	if err != nil {
+		return fmt.Errorf("failed to get block header "+
+			"metadata: %w", err)
+	}
+
+	totalHeaders := endHeight - startHeight + 1
+	log.Infof("Appending %d new headers in batches of %d", totalHeaders,
+		s.options.WriteBatchSizePerRegion)
+
+	// Process headers in batches.
+	for batchStart := startHeight; batchStart <= endHeight; {
+		// Calculate batch end, ensuring we don't exceed endHeight.
+		batchEnd := min(
+			batchStart+uint32(s.options.WriteBatchSizePerRegion)-1,
+			endHeight,
+		)
+
+		// Calculate indices in the source based on the metadata start
+		// height.
+		sourceStartIdx := targetHeightToImportSourceIndex(
+			batchStart, metadata.StartHeight,
+		)
+		sourceEndIdx := targetHeightToImportSourceIndex(
+			batchEnd, metadata.StartHeight,
+		)
+
+		// Read block headers batch.
+		blockHeaders := make(
+			[]headerfs.BlockHeader, 0, batchEnd-batchStart+1,
+		)
+		blockIter := s.BlockHeadersImportSource.Iterator(
+			sourceStartIdx, sourceEndIdx,
+		)
+		defer blockIter.Close()
+
+		for {
+			header, hasMore, err := blockIter.Next()
```

Comment:
Can use that batch iteration w/ `iter.Seq` once again. 

I think the approach of validating them all at once, then writing wors, at least when we have the `mmap` version. For teh HTTP variant, we may want to reconsider that, as if the network fails, then we'll need to validate them all over again as we don't have them on disk.